src/main/java/com/streamr/client/Subscription.java

```diff
@@ -49,4 +73,96 @@ public void setState(State state) {
     public MessageHandler getHandler() {
         return handler;
     }
+
+    public boolean isResending() {
+        return resending;
+    }
+
+    public void setResending(boolean resending) {
+        this.resending = resending;
+    }
+
+    public ResendRangeRequest handleMessage(StreamMessage msg, String sessionToken) {
+        return handleMessage(msg, sessionToken, false);
+    }
+
+    public ResendRangeRequest handleMessage(StreamMessage msg, String sessionToken, boolean isResend) {
+        String key = msg.getPublisherId() + msg.getMsgChainId();
+        if (resending && !isResend) {
+            queue.add(msg);
+            return null;
+        } else if (checkForGap(msg.getPreviousMessageRef(), key) && !resending) {
+            queue.add(msg);
+            MessageRef from = lastReceivedMsgRef.get(key); // cannot know the first missing message so there will be a duplicate received
```

Comment:
The next message could also be with a timestamp strictly larger than the last received (we don't know by how much) and sequence number 0. That's why when we discussed this issue for the Javascript client we implemented it with the first message being a duplicate.
This is because we compute the gap on the recipient side. In TCP, the recipient just sends back the last sequence number received and from that information the sender can compute the exact gap because the sender knows the sequence of messages.

Reply:
But won't the server essentially return messages which are:

`(ts = requestedTs AND seq >= requestedSeq) OR ts > requestedTs`?

So if we request messages from `(lastSeenTs, lastSeenSeq + 1)`, it should correctly give us what we want (any further messages for the same timestamp, and any messages with larger timestamp), but not the `lastSeen` message again?

In the our message ordering, `(ts, seq+1)` is the closest possible next message to `(ts, seq)`, in that there can be no messages in between them. These would match the first clause before `OR`. 

All `(ts+, *)` are strictly larger than `(ts, *)` when `ts+ > ts`. These would match the second clause after `OR`.

Reply:
But won't the server essentially return messages which are:

`(ts = requestedTs AND seq >= requestedSeq) OR ts > requestedTs`?

So if we request messages from `(lastSeenTs, lastSeenSeq + 1)`, it should correctly give us what we want (any further messages for the same timestamp, and any messages with larger timestamp), but not the `lastSeen` message again?

In the our message ordering, `(ts, seq+1)` is the closest possible next message to `(ts, seq)`, in that there can be no messages in between them. These would match the first clause before `OR`. 

All `(ts+, *)` are strictly larger than `(ts, *)` when `ts+ > ts`. These would match the second clause after `OR`.